### PR TITLE
Add Keras 3 support to ParametricUMAP

### PR DESCRIPTION
**Pull request by fchollet**
I've tested basic use cases but there may be more needed. Also, docs may need to be updated as well.

This makes the code work with both Keras 2 and Keras 3 with the TF backend.

### CODE REVIEW

**Comment by fchollet** [comment]
I've fixed most of the test failures, but unfortunately I have no idea what the "non parametric" case represents and how it works, so I'm not able to fix that one.

**Comment by lmcinnes** [comment]
Thanks for this. I will try to make some time tomorrow to see if I can work out the errors here.

**Comment by timsainb** [comment]
I haven’t looked at the code (I can if you’d like), just wanted to chime in - Non parametric is regular UMAP but with gradient descent done in tensorflow. I don’t think it has any real use in the python package, it was just to for comparison in the paper.

**Comment by fchollet** [comment]
> I don’t think it has any real use in the python package, it was just to for comparison in the paper.

Got it, thanks for the explanation. The way it's implemented seems very unusual. My recommendation is still to remove that part of the code, to make the library simpler and more maintainable. It will also make it easier to make the code compatible with all backends in the future.

**Comment by fchollet** [comment]
> It will also make it easier to make the code compatible with all backends in the future.

When we do that there are a few more refactorings we should implement.

1. It would be nicer to make a `Model` subclass that owns the encoder and decoder, and chains the two in `call()`. With a custom `compute_loss()` method. Easier to manage than the multiple Functional models we have now (which include a `Lambda` layer, which isn't recommended).

2. The gradient clipping in the `train_step()` isn't necessary because the optimizer can do the same. It's better not to have a custom `train_step()`, because that would force us to write 3 different implementations to achieve compatibility with JAX/torch.

3. Loss computation seems like it could be simplified.

**Comment by lmcinnes** [comment]
I agree, we should just remove the non-parametric. The other errors are all due to a change in pytest that I can (and already have but haven't pushed yet) clean up. I can merge this now if you like and handle the clean up (including removal of non-parametric) myself.

As to further refactoring -- now that @timsainb is here I will leave that to a discussion between both of you; this definitely is not my area of expertise.

**Comment by fchollet** [comment]
> I can merge this now if you like and handle the clean up (including removal of non-parametric) myself.

Sure, that sounds good to me. Thank you.

> now that @timsainb is here I will leave that to a discussion between both of you; this definitely is not my area of expertise.

Sure. @timsainb Tim, what do you think about the refactorings I suggested?

**Comment by timsainb** [comment]
Hi @fchollet thanks for contributing this! Seems like not so long ago I was switching from lasagne to keras 1 :)

> It would be nicer to make a Model subclass that owns the encoder and decoder, and chains the two in call(). With a custom compute_loss() method. Easier to manage than the multiple Functional models we have now (which include a Lambda layer, which isn't recommended).

This makes sense. 

> The gradient clipping in the train_step() isn't necessary because the optimizer can do the same. It's better not to have a custom train_step(), because that would force us to write 3 different implementations to achieve compatibility with JAX/torch.

Moving gradient clipping to the optimizer sounds good.

 I'm a little confused about the JAX/torch compatibility, what is the idea here? IMO it would be great to move to keras only and allow for jax/torch in the backend. 

> Loss computation seems like it could be simplified.

Interested to see how!

**Comment by fchollet** [comment]
> IMO it would be great to move to keras only and allow for jax/torch in the backend.

Yes, that is the idea. The code would be written with the `keras.ops` API and would be able to run on top of JAX/torch by switching to a different backend.

I'm thinking we could have a model like this:

```python

class UMAPModel(keras.Model):
    def __init__(self,
                 encoder,
                 decoder,
                 optimizer=None,
                 parametric_reconstruction=False,
                 parametric_reconstruction_loss_weight=1.,
                 global_correlation_loss_weight=0.,
                 name="umap_model"):
        super().__init__(name=name)
        self.encoder = encoder
        self.decoder = decoder
        self.parametric_reconstruction = parametric_reconstruction
        self.global_correlation_loss_weight = global_correlation_loss_weight
        self.parametric_reconstruction_loss_weight = parametric_reconstruction_loss_weight
        
        self._total_loss_tracker = keras.metrics.Mean(name="loss")
        optimizer = optimizer or keras.optimizers.Adam(1e-3, clipvalue=4.0)
        self.compile(optimizer=optimizer)

    def call(self, inputs):
        to_x, from_x = inputs
        embedding_to = self.encoder(to_x)
        embedding_from = self.encoder(from_x)

        y_pred = {
            "embedding_to": embedding_to,
            "embedding_from": embedding_from,
        }
        if self.parametric_reconstruction:
            # parametric reconstruction
            if self.autoencoder_loss:
                embedding_to_recon = self.decoder(embedding_to)
            else:
                # stop gradient of reconstruction loss before it reaches the encoder
                embedding_to_recon = self.decoder(ops.stop_gradient(embedding_to))
            y_pred["reconstruction"] = embedding_to_recon
        return y_pred

    def compute_loss(
        self, x=None, y=None, y_pred=None, sample_weight=None, allow_empty=False
    ):
        losses = []
        # Regularization losses.
        for loss in self.losses:
            losses.append(tf.cast(loss, dtype=keras.backend.floatx()))

        # umap loss
        losses.append(self._umap_loss(y, y_pred))

        # global correlation loss
        if self.global_correlation_loss_weight > 0:
            losses.append(self._global_correlation_loss(y, y_pred))

        # parametric reconstruction loss
        if self.parametric_reconstruction:
            losses.append(self._parametric_reconstruction_loss(y, y_pred))

        return ops.sum(losses)
    
    def _umap_loss(self, y, y_pred):
        # TODO
        return loss

    def _global_correlation_loss(self, y, y_pred):
        # TODO
        return loss * self.global_correlation_loss_weight

    def _parametric_reconstruction_loss(self, y, y_pred):
        # TODO
        return loss * self.parametric_reconstruction_loss_weight
```

Does that make sense?

Also, are all the constructor arguments in ParametricUMAP necessary? There's a lot going on, I wonder how often these actually get used.

**Comment by timsainb** [comment]
Yeah that looks a lot cleaner. 
 
Here are the constructor arguments that could be considered for the chopping block: 
- `run_eagerly`: I can't think off the top of my head why it would be valuable to run eagerly but I imagine that would be something handled by keras. 
- `loss_report_frequency` and `n_training_epochs`: this is a bit wonky and probably most people don't modify either parameter. Right now, `n_training_epochs` defaults to 1 (i.e. train for a single epoch) and `loss_report_frequency' defaults to 10, so that you can see the loss go down over the course of that single epoch. 
   - Understanding these parameters requires explaining how epochs work in Parametric UMAP, which is a bit of a weird thing:
      - In UMAP, there are by default 200 epochs. Not all edges are sampled/trained on in each epoch, and whether an edge is sampled depends on the edge probability (i.e. a probability of 1 is trained for all 200 epochs). @lmcinnes its been a while, correct me if I'm wrong.
      - In Parametric UMAP, rather than create a new dataset for each epoch, I pre-generated the tf.data.Dataset of edges (with repeats set based on the probabilities) for all 200 epochs as a single dataset, so 1 epoch of Parametric UMAP = all 200 epochs of regular UMAP. 
      -  we treat each of these 200 epochs as a single training epoch, and the edge dataset is constructed over the indices of the graph, which are differentially sampled based on the edge probability (see https://github.com/lmcinnes/umap/blob/868e55cb614f361a0d31540c1f4a4b175136025c/umap/parametric_umap.py#L868). Then when a batch of edges is fetched it gets the corresponding input data for each edge with tf.gather. 
   - in summary, 
      - 'n_epochs' is a constructor argument that refers to the maximum number of times that an edge is trained on for both UMAP and Parametric UMAP. 
      -  'n_training_epochs' is specific to Parametric UMAP, and refers to the number of epochs you want to repeat the 200 UMAP training epochs (which is usually just once). 
      - `loss_report_frequency` accounts for that weirdness by allowing you to report the loss mid `n_training_epochs` epoch. 
   - A potentially cleaner alternative is to generate a new 'dataset' of edges for each of the 200 UMAP training epochs. I imagine this would be slower to train though (which I believe is why I set it up the way that I did). 
- `reconstruction_validation`: maybe fewer people are using this, the main purpose is early stopping, specific to the reconstruction error. It also would be valuable to have the same thing for early stopping UMAP but this was never implemented. 
- `optimizer`: Generally ADAM works fine

These are important:
- `batch_size`: probably the default is fine and used by most people
- `dims`: shape of the data. necessary for dealing with non-flat data (e.g. with a convnet encoder). 
- `encoder` and `decoder`: for using your own models (e.g. a convnet) 
- `parametric_reconstruction`: for better and faster reconstruction than using nearest neighbors
- `parametric_reconstruction_loss_fcn`: we default to BinaryCrossentropy but you can imagine for non binary data you would want to use SSE (even though it would be harder to balance out the umap and reconstruction loss, similar to needing beta in a vae) 
- `parametric_reconstruction_loss_weight` necessary if the scale of the reconstruction error is much higher or lower than umap loss
- - `global_correlation_loss_weight`: important for being able to add more global structure to embedding (like balancing UMAP and MDS) 
- `autoencoder_loss`: (True/False) I don't know how often people are jointly the embedder Parametric UMAP with autoencoder loss but I think its valuable.

Also found a few vestiges of nonparametric umap that can be updated: 

- we can get rid of `make_sham_generator` now that non-parametric embeddings are gone: https://github.com/lmcinnes/umap/blob/868e55cb614f361a0d31540c1f4a4b175136025c/umap/parametric_umap.py#L845
-  optimizer argument can default to adam, before we had different optimizers for parametric and nonparametric: https://github.com/lmcinnes/umap/blob/868e55cb614f361a0d31540c1f4a4b175136025c/umap/parametric_umap.py#L145

**Comment by fchollet** [comment]
Cool, I guess we could cut out the arguments you mentioned as non-essential and keep the others.

How of this are you willing to do (model refactoring + cleanups + making it backend-agnostic)? How much do you want me to do?
